files/default/neutron-ha-tool.py

```diff
@@ -576,6 +600,199 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+    :param wait_for_network: Optional wait-for-dhcp-network flag
+    :param fail_fast: Optional fail-on-first-error flag
+    :returns: Total number of errors encountered
+
+    """
+
+    agent_list = list_agents(qclient, agent_type='DHCP agent')
+
+    agent_to_evacuate = None
+    for agent in agent_list:
+        if agent.get('host', None) == agent_host:
+            agent_to_evacuate = agent
+            break
+
+    if not agent_to_evacuate:
+        LOG.error("Could not locate agent to evacuate; aborting!")
+        return 0
+
+    target_list = list_alive_dhcp_agents_except(agent_list,
+                                                exclude_agent=agent_to_evacuate)
+
+    if len(target_list) < 1:
+        LOG.error("There are no DHCP agents alive to migrate networks onto")
+        return 0
+
+    (migrations, errors) = \
+        migrate_dhcp_networks_from_agent(qclient, agent_to_evacuate,
+                                      target_list, agent_picker,
+                                      noop, wait_for_network, fail_fast)
+    LOG.info("%d networks %s evacuated from DHCP agent %s", migrations,
+             "would have been" if noop else "were", agent_host)
+    if errors > 0:
+        LOG.error("%d errors encountered during evacuation", errors)
+
+    return errors
+
+
+def migrate_dhcp_networks_from_agent(qclient, agent, targets, agent_picker,
+                                     noop, wait_for_network, fail_fast):
+    LOG.info("Querying agent_id=%s for networks to migrate away", agent['id'])
+
+    networks_on_agent = qclient.list_networks_on_dhcp_agent(agent['id'])['networks']
+
+    agent_picker.set_agents(targets)
+
+    migrations = 0
+    errors = 0
+    for network in networks_on_agent:
+        target = agent_picker.pick()
+
+        if noop:
+            LOG.info("Would try to migrate network=%s from agent=%s "
+                 "to agent=%s", network['id'], agent['id'], target['id'])
+            migrations += 1
+            continue
+
+        try:
+            migrate_dhcp_network(qclient, network, agent, target, wait_for_network)
+            migrations += 1
+        except:
+            LOG.exception("Failed to migrate network_id=%s from agent=%s to "
+                          "agent=%s", network['id'], agent['id'], target['id'])
+            errors += 1
+            if fail_fast:
+                break
+
+    return (migrations, errors)
+
+
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
+
+    :param qclient: A neutronclient
+    :param network: The network to migrate
+    :param agent: The DHCP agent to migrate from
+    :param target: The DHCP agent to migrate to
+    :param wait_for_network: Optional wait-for-dhcp-network flag
+    """
+
+    LOG.info("Migrating network=%s from agent=%s to agent=%s",
+             network['id'], agent['id'], target['id'])
+
```

Comment:
`network['id'], agent['id'], target['id']` are used several times in this function, maybe save them into variables?

Reply:
done.